### PR TITLE
Add simple test scopes for checking type inference failures

### PR DESCRIPTION
**Pull request by Xerios**
I ported back a simple version of `test optionalName { ... }` scopes, including an example file. 

Test declaration is simply a copy of module as its behavior requirements fit well, which is why it's called TestModule. 
When name is not set, we generate a new test_randomId name. It can also be set to "FAIL" (`test FAIL { .. }`) which passes when infer fails parsing the code.

Failed tests do not print code as we can no longer guarantee types being present.

**In my implementation,** the surrounding comments (which I left in, but commented out) are used to create individual tests, which keeps all the surrounding non-test-scoped code but only one includes one test declaration. So a file with multiple tests would generate and compile multiple versions of the same source and evaluate them one by one. Test comments also include source location so that the test runner doesn't have to interact with the compiler.

Sorry about the formatting, my default formatter is quite different so I had to disable it and manually adjust things.

### CODE REVIEW

**Comment by nathsou** [comment]
Hi,

Again thanks for your work, I have a few issues with the current design:

1. You can specify that a test block should fail but there is no way to check the actual error so it might fail for a different reason than what you're expecting, I imagined to handle this in a similar way to https://github.com/munificent/craftinginterpreters/blob/master/test/assignment/infix_operator.lox#L3

2. I'm not a huge fan of the `test FAIL {...}` syntax, I think we should be able to make positive and negative assertions/expectations in the same test block and it's quite strange to me to allow any identifier for the name of tests, but then one particular string "FAIL" behaves completely differently.

3. This is not super relevant for this task as it doesn't actually run tests but I haven't yet made up my mind on wether test files should always be separate from the main code and always run or if they can be included in any file and then run conditionally if a test command is run, I still have to think about it.

So to sum up, I really haven't given test blocks much thoughts yet as I'm focusing on other areas of the project and I am also quite busy with other stuff so if you want to address some of these issues and propose some design ideas, go for it. Otherwise you can enjoy this implementation in you own fork right now :)

**Comment by Xerios** [comment]
Well, it is a very simple implementation. A lazy one too because not a lot of thought went into it.

Unfortunately there are no one-size-fits-all solution when it comes to how and where tests are done. Is the "FAIL" syntax crap and very uninspiring? Definitely, but it is very simple and achieves its only job. 

> Is the "FAIL" syntax crap and uninspiring? 

Definitely, but it is very simple and achieves its only job. I haven't found an actual use for identifiers yet, it was simply due to me replicating module logic. I think having a label would fit better if the need ever arises. 

In my case, inline `pass/fail tests` work for my current flow. I doubt that such simple test logic would be capable handling more advanced scenarios in the future, it will need to evolve eventually. 

_However_, at early stage, something dumb and simple allows me to easily write test on-the-go and collect all failure cases which I will need to eventually handle (as todos) or provide more descriptive errors for mistakes that aren't that obvious. 

My actual solution is further developed and is closer to "crafting interpreters" example, although I'm following the TS-style errors where errors have a prefix (defining the phase e.g. parser/types/linter/general) with an ID, the verbose description is more of a comment, and the error location is optional. Goal is so that I don't have to revisit and re-adapt any old tests often. **Bit of a fire-and-forget approach.**

It's understandable why this isn't a priority for you, after all you have yet to fully design Poy. In my case the syntax and behavior is already well-defined, but I'm not as well-versed in compiler design as you are so I happen to make more mistakes. Those mistakes that are quickly caught by these simple test cases I've been building up. I'm just dum-dum proofing for myself 😅
